packages/components/src/handler.ts

```diff
@@ -178,9 +180,30 @@ export class CustomChainHandler extends BaseCallbackHandler {
         this.socketIO.to(this.socketIOClientId).emit('end')
     }
 
-    handleChainEnd(outputs: ChainValues): void | Promise<void> {
-        if (this.returnSourceDocuments) {
-            this.socketIO.to(this.socketIOClientId).emit('sourceDocuments', outputs?.sourceDocuments)
+    handleChainEnd(outputs: ChainValues, _: string, parentRunId?: string): void | Promise<void> {
+        /*
+            Langchain does not call handleLLMStart, handleLLMEnd, handleLLMNewToken when the chain is cached.
+            Callback Order is "Chain Start -> LLM Start --> LLM Token --> LLM End -> Chain End" for normal responses.
+            Callback Order is "Chain Start -> Chain End" for cached responses.
+         */
+        if (this.cachedResponse && parentRunId === undefined) {
```

Comment:
found out that if used in ConversationRetrievalQAChain where there are multiple chains, this get triggered multiple times. Adding check for `parentRunId` to keep track what is the last chain, and only emit token for that